x-pack/metricbeat/module/openai/usage/persistcache.go

```diff
@@ -0,0 +1,172 @@
+	"os"
+	"path"
+	"strings"
+	"sync"
+	"time"
+)
+
+// stateManager handles the storage and retrieval of state data with key hashing and caching capabilities
+type stateManager struct {
+	store     *stateStore
+	keyPrefix string
+	hashCache sync.Map
+}
+
+// stateStore handles persistence of key-value pairs using the filesystem
+type stateStore struct {
+	sync.RWMutex        // Protects access to the state store
+	Dir          string // Base directory for storing state files
+}
+
+// newStateManager creates a new state manager instance with the given storage path
+func newStateManager(storePath string) (*stateManager, error) {
+	if strings.TrimSpace(storePath) == "" {
+		return nil, errors.New("empty path provided")
+	}
+
+	store, err := newStateStore(storePath)
+	if err != nil {
+		return nil, fmt.Errorf("create state store: %w", err)
+	}
+
+	return &stateManager{
+		store:     store,
+		keyPrefix: "state_",
+		hashCache: sync.Map{},
+	}, nil
+}
+
+// newStateStore creates a new state store instance at the specified path
+func newStateStore(path string) (*stateStore, error) {
+	if err := os.MkdirAll(path, 0o755); err != nil {
+		return nil, fmt.Errorf("creating state directory: %w", err)
+	}
+	return &stateStore{
+		Dir: path,
+	}, nil
+}
+
+// getStatePath builds the full file path for a given state key
+func (s *stateStore) getStatePath(name string) string {
+	return path.Join(s.Dir, name)
+}
+
+// Put stores a value in a file named by the key
+func (s *stateStore) Put(key string, value string) error {
+	s.Lock()
+	defer s.Unlock()
+
+	filePath := s.getStatePath(key)
+
+	// In case the file already exists, file is truncated.
+	f, err := os.Create(filePath)
+	if err != nil {
+		return fmt.Errorf("creating state file: %w", err)
+	}
+	defer f.Close()
+
+	_, err = f.WriteString(value)
+	if err != nil {
+		return fmt.Errorf("writing value to state file: %w", err)
+	}
+
+	if err = f.Sync(); err != nil {
+		return fmt.Errorf("syncing state file: %w", err)
+	}
+
+	return nil
+}
+
+// Get retrieves the value stored in the file named by the key
+func (s *stateStore) Get(key string) (string, error) {
+	s.RLock()
+	defer s.RUnlock()
+
+	filePath := s.getStatePath(key)
+	data, err := os.ReadFile(filePath)
+	if err != nil {
+		return "", fmt.Errorf("reading state file: %w", err)
+	}
+	return string(data), nil
+}
+
+// Has checks if a state exists for the given key
+func (s *stateStore) Has(key string) bool {
+	s.RLock()
+	defer s.RUnlock()
+
+	filePath := s.getStatePath(key)
+	_, err := os.Stat(filePath)
+	return err == nil
+}
+
+// Remove deletes the state file for the given key
+func (s *stateStore) Remove(key string) error {
+	s.Lock()
+	defer s.Unlock()
+
+	filePath := s.getStatePath(key)
+	if err := os.Remove(filePath); err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("removing state file: %w", err)
+	}
+	return nil
+}
+
+// Clear removes all state files by deleting and recreating the state directory
+func (s *stateStore) Clear() error {
```

Comment:
If our intent here is to remove the files, since we are again recreating the directory. Should we delete just the files instead?
It might avoid potential permission issues or race conditions. WDYT ?



Eg:

    files, err := os.ReadDir(s.Dir)
    if err != nil {
        return fmt.Errorf("reading state directory: %w", err)
    }

    for _, file := range files {
        if err := os.Remove(path.Join(s.Dir, file.Name())); err != nil {
            return fmt.Errorf("removing state file: %w", err)
        }
    }

Reply:
Actually, when I was writing this I was creating it to be good package for reuse later as well and this method was inteded to completely reset the state. Plan earlier was to add subdirectories too inside the state folder because what if in future we add a new dataset. For now, we just have "usage", but in case we add more.

But as it just resides in this package, we are not even using this method. It is unused. We can even remove this method.

So, that's why this. It's a complete reset of state.